src/tools/rbd_mirror/image_map/Types.h

```diff
@@ -0,0 +1,97 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+
+#include <map>
+#include <string>
+#include <boost/variant.hpp>
+
+#include "include/buffer.h"
+#include "include/encoding.h"
+#include "tools/rbd_mirror/types.h"
+
+namespace ceph {
+class Formatter;
+}
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+enum MirrorPolicyMetaType {
+  MIRROR_POLICY_NONE_TYPE = 0,
+};
+
+struct MirrorPolicyMetaNone {
+  static const MirrorPolicyMetaType TYPE = MIRROR_POLICY_NONE_TYPE;
+
+  MirrorPolicyMetaNone() {
+  }
+
+  void encode(bufferlist& bl) const {
+  }
+
+  void decode(__u8 version, bufferlist::iterator& it) {
+  }
+
+  void dump(Formatter *f) const {
+  }
+};
+
+struct MirrorPolicyMetaUnknown {
+  static const MirrorPolicyMetaType TYPE = static_cast<MirrorPolicyMetaType>(-1);
+
+  MirrorPolicyMetaUnknown() {
+  }
+
+  void encode(bufferlist& bl) const {
+    assert(false);
+  }
+
+  void decode(__u8 version, bufferlist::iterator& it) {
+  }
+
+  void dump(Formatter *f) const {
+  }
+};
+
+typedef boost::variant<MirrorPolicyMetaNone,
+                       MirrorPolicyMetaUnknown> MirrorPolicyMeta;
+
+struct MirrorPolicyData {
+  MirrorPolicyData()
+    : policy_meta(MirrorPolicyMetaNone()) {
+  }
+  MirrorPolicyData(MirrorPolicyMeta policy_meta)
+    : policy_meta(policy_meta) {
+  }
+
+  MirrorPolicyMeta policy_meta;
+
+  MirrorPolicyMetaType get_policy_meta_type() const;
+
+  void encode(bufferlist& bl) const;
+  void decode(bufferlist::iterator& it);
+  void dump(Formatter *f) const;
+
+  static void generate_test_instances(std::list<MirrorPolicyData *> &o);
+};
+
+enum ShuffleType {
+  SHUFFLE_TYPE_INSTANCES_ADDED = 0,   // rebalance due to addition of instances
+  SHUFFLE_TYPE_INSTANCES_REMOVED,     // rebalance due to removal of instances
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
```

Comment:
Nit: case doesn't appear to be used (of course, not sure the ```ShuffleType``` would need to live either)